### PR TITLE
[no-119 (iohk-monitoring-framework)] Logging in Sim.

### PR DESCRIPTION
**Pull request by 365andreas**
https://github.com/input-output-hk/iohk-monitoring-framework/issues/119

@coot Could you also suggest a better place to put `LogTrace` ?

### CODE REVIEW

**Comment by coot** [comment]
What about `Control.Monad.Logger` or something similar like `Control.Monad.Logger.Trace`.  The former would clash with some libraries, so the latter is better.

File: io-sim/src/Control/Monad/IOSim.hs [hunk 1]
```diff
@@ -9,6 +9,7 @@
 {-# LANGUAGE ScopedTypeVariables        #-}
 {-# LANGUAGE TypeFamilies               #-}
 {-# OPTIONS_GHC -Wno-orphans            #-}
+{-# LANGUAGE AllowAmbiguousTypes        #-}
```
Comment (dcoutts) [inline io-sim/src/Control/Monad/IOSim.hs]: Eeek!

File: io-sim-classes/src/Control/Monad/Class/MonadTimer.hs [hunk 2]
```diff
@@ -42,7 +42,7 @@ fromStart = flip addTime zero
 
 class (Monad m, TimeMeasure (Time m), Show (Time m)) => MonadTime m where
   type Time m :: *
-
+  toInt :: Time m -> m Int
```
Comment (dcoutts) [inline io-sim-classes/src/Control/Monad/Class/MonadTimer.hs]: I've been thinking we should make time durations be a fixed type, not something determined by the monad. This might eliminate the need for this (rather under-specified) operation.

File: io-sim-classes/src/Control/Monad/Class/MonadProbe.hs [hunk 3]
```diff
@@ -60,3 +82,150 @@ instance MonadRunProbe IO IO where
   newProbe  = ProbeIO <$> STM.newTVarIO []
   readProbe (ProbeIO p) = STM.readTVarIO p
   runM = id
+
+
+data LogTrace t' m t = LogTrace {
+      traceOutput :: t -> m ()
+    , readLogTrace :: m t'
+    , traceStopping :: m ()
```
Comment (dcoutts) [inline io-sim-classes/src/Control/Monad/Class/MonadProbe.hs]: What are the second two for? I thought a tracer was just an output mechanism not a query mechanism?

And what is the meaning of the `t'` param?

File: io-sim-classes/src/Control/Monad/Class/MonadProbe.hs [hunk 4]
```diff
@@ -60,3 +82,150 @@ instance MonadRunProbe IO IO where
...
+                                , liSeverity  = Info
+                                , liPayload   = pack a
+                                }
+                atomically $ modifyTVar' c ((LogObject meta content):)
+            , readLogTrace = atomically $ do
+                v <- readTVar c
+                writeTVar c []
+                return $ reverse v
+            , traceStopping = fail "Not supported operation"
+            }
```
Comment (dcoutts) [inline io-sim-classes/src/Control/Monad/Class/MonadProbe.hs]: I'm not clear why all these utils live here in this package. Many of them are `MonadIO` and nothing to do with the classes in this package.

File: io-sim/src/Control/Monad/IOSim.hs [hunk 5]
```diff
@@ -191,6 +195,9 @@ instance MonadSay (SimM s) where
   say msg = SimM $ \k -> Say msg (k ())
 
 instance MonadFork (SimM s) where
+  type ThreadIdT (SimM s) = ThreadId
```
Comment (dcoutts) [inline io-sim/src/Control/Monad/IOSim.hs]: I think we'll be able to call this `ThreadId` rather than `ThreadIdT`. See how we resolve the name clashes with `TVars` etc.

File: io-sim/src/Control/Monad/IOSim.hs [hunk 6]
```diff
@@ -735,31 +746,31 @@ execAtomically mytid = go [] []
   where
     go :: [SomeTVar s] -> [SomeTVar s] -> TVarId
        -> StmA s a -> ST s (StmTxResult s a)
-    go read written nextVid action = case action of
-      ReturnStm x     -> do (vids, tids) <- finaliseCommit written
+    go read_tvs written_tvs nextVid action = case action of
```
Comment (dcoutts) [inline io-sim/src/Control/Monad/IOSim.hs]: What's all this renaming for? We can hide imports to avoid name clashes.

File: typed-protocols/src/Network/TypedProtocol/Driver.hs [hunk 7]
```diff
@@ -68,27 +69,30 @@ import Numeric.Natural (Natural)
...
           Maybe bytes
        -> Peer ps pr st' m a
        -> m (Either failure a)
-    go trailing (Effect k) = k >>= go trailing
-    go _        (Done _ x) = return (Right x)
+    go trailing (Effect k) = traceOutput tr "Effect" >> k >>= go trailing
+    go _        (Done _ x) = traceOutput tr "Done"   >>return (Right x)
 
     go trailing (Yield stok msg k) = do
+      traceOutput tr "Yield"
```
Comment (coot) [inline typed-protocols/src/Network/TypedProtocol/Driver.hs]: it should be possible to do `traceOutput tr ("Yield " ++ show msg)`, this will be more useful.

File: typed-protocols/src/Network/TypedProtocol/Driver.hs [hunk 8]
```diff
@@ -68,27 +69,30 @@ import Numeric.Natural (Natural)
...
       send (encode stok msg)
       go trailing k
 
     go trailing (Await stok k) = do
+      traceOutput tr "Await"
```
Comment (coot) [inline typed-protocols/src/Network/TypedProtocol/Driver.hs]: You could also add modify `k` so that it logs when a message is received.

File: typed-protocols/src/Network/TypedProtocol/PingPong/Tests.hs [hunk 9]
```diff
@@ -310,13 +316,27 @@ prop_channel (NonNegative n) = do
...
-    ioProperty (prop_channel n)
-
-prop_channel_ST :: NonNegative Int -> Bool
-prop_channel_ST n = 
-    case runSim (prop_channel n) of
-      Right True -> True
-      _          -> False
+    ioProperty $ do
+        startTime <- (,) <$> getCurrentTime <*> getMonotonicTime
+        -- prop <- withLoggingToMonitoringFramework (prop_channel n)
```
Comment (coot) [inline typed-protocols/src/Network/TypedProtocol/PingPong/Tests.hs]: maybe this shouldn't be commented out

File: io-sim/src/Control/Monad/IOSim.hs (hunk 6 above)
Comment (CodiePP) [reply io-sim/src/Control/Monad/IOSim.hs]: I proposed this as it is quite ugly to have a variable named the same as some function in Prelude. And, in this case the name `read` is not very indicative.
The alternative is to import the Prelude's read function again in a qualified import statement. More ugliness!

File: typed-protocols/src/Network/TypedProtocol/Driver.hs (hunk 7 above)
Comment (365andreas) [reply typed-protocols/src/Network/TypedProtocol/Driver.hs]: `msg` is `Message ps st st` and it is not `Show`able
so added some `Show bytes` constraints in order to trace the outcome of `(encode stok msg)`.

Comment (coot) [reply typed-protocols/src/Network/TypedProtocol/Driver.hs]: Yes that's right, but you could use `Network.TypedProtocol.Codec.AnyMessage` to wrap `msg` and then use `show (AnyMessage msg)`.

Comment (365andreas) [reply typed-protocols/src/Network/TypedProtocol/Driver.hs]: It is not `Show`able either.
` Could not deduce (Show (AnyMessage ps))`

Comment (coot) [reply typed-protocols/src/Network/TypedProtocol/Driver.hs]: That's true, but you can add it as a constraint, no?

File: io-sim-classes/src/Control/Monad/Class/MonadFork.hs [hunk 10]
```diff
@@ -11,14 +12,22 @@ import           Control.Monad.Reader
...
+import           Cardano.BM.Data.LogItem (ThreadIdInt)
+
 forkPrintExceptionLock :: IO.MVar ()
 {-# NOINLINE forkPrintExceptionLock #-}
 forkPrintExceptionLock = unsafePerformIO $ IO.newMVar ()
 
 class Monad m => MonadFork m where
+  type ThreadIdT m :: *
+  myThreadId :: m (ThreadIdT m)
+  toThreadIdInt :: (ThreadIdT m) -> m ThreadIdInt
```
[the remark this answers is not in the document]
Comment (365andreas) [reply io-sim-classes/src/Control/Monad/Class/MonadFork.hs]: Do we need a new class for this?

File: typed-protocols/src/Network/TypedProtocol/Driver.hs (hunk 7 above)
Comment (coot) [reply typed-protocols/src/Network/TypedProtocol/Driver.hs]: You're here polymorphic in the protocol (the `ps` parameter), so in your framework you cannot have it satisfied, but you can assume it.  If it is used somewhere, you might get some errors that `Show (AnyMessage PingPong)` does not have `Show` instance, but there's no problem with adding one.

Comment (coot) [reply typed-protocols/src/Network/TypedProtocol/Driver.hs]: Just be careful so that we're not left with an orphaned instance.

File: io-sim-classes/src/Control/Monad/Class/MonadProbe.hs (hunk 4 above)
Comment (365andreas) [reply io-sim-classes/src/Control/Monad/Class/MonadProbe.hs]: > @coot Could you also suggest a better place to put `LogTrace` ?

> What about `Control.Monad.Logger` or something similar like `Control.Monad.Logger.Trace`. The former would clash with some libraries, so the latter is better.
https://github.com/input-output-hk/ouroboros-network/pull/298#issuecomment-464067480

Comment (coot) [reply io-sim-classes/src/Control/Monad/Class/MonadProbe.hs]: That's true; just use the `ouroboros-network` package for `Control.Monad.Logger`. This way any upstream dependency will have access to it. If you prefer it to be independent of `ouroboros-network` then you could add another package for that. That might be useful if at some point one will want to use `io-sim` and the logger without requiring `ouroboros-network`.

File: io-sim-classes/src/Control/Monad/Class/MonadFork.hs (hunk 10 above)
[the remark this answers is not in the document]
Comment (coot) [reply io-sim-classes/src/Control/Monad/Class/MonadFork.hs]: I don't think so, you just need one when you create a `LogTrace`, that can be local to the module or definition of a `LogTrace`.

File: typed-protocols/src/Network/TypedProtocol/Driver.hs (hunk 7 above)
Comment (365andreas) [reply typed-protocols/src/Network/TypedProtocol/Driver.hs]: > That's true, but you can add it as a constraint, no?
I think yes with the same way I added `Show bytes` constraints.

Comment (coot) [reply typed-protocols/src/Network/TypedProtocol/Driver.hs]: Indeed :)

File: io-sim-classes/src/Control/Monad/Class/MonadFork.hs (hunk 10 above)
[the remark this answers is not in the document]
Comment (365andreas) [reply io-sim-classes/src/Control/Monad/Class/MonadFork.hs]: How can this be polymorphic on the monad so as to have `myThreadId :: m ThreadId` in both `IO` and `SimM`?

File: io-sim-classes/src/Control/Monad/Class/MonadProbe.hs (hunk 3 above)
Comment (365andreas) [reply io-sim-classes/src/Control/Monad/Class/MonadProbe.hs]: In `SimM`: `traceOutput` will write in a `TVar`and at the end `readLogTrace` will read from it.
Also, `iohk-monitoring-framework`'s `Trace` needs a shutdown.

`t'` is the type of data saved in the `TVar`. To be more specific a msg could be "hello" but we want also to keep the timestamp and the threadId of the sender so `t'` could be `t' :: (ThreadId, Timestamp, String)`.

File: io-sim-classes/src/Control/Monad/Class/MonadFork.hs (hunk 10 above)
[the remark this answers is not in the document]
Comment (coot) [reply io-sim-classes/src/Control/Monad/Class/MonadFork.hs]: I see; what are you using this function for, just for showing the `ThreadId`?

[the remark this answers is not in the document]
Comment (365andreas) [reply io-sim-classes/src/Control/Monad/Class/MonadFork.hs]: Yes exactly.

File: io-sim/src/Control/Monad/IOSim.hs (hunk 6 above)
Comment (dcoutts) [reply io-sim/src/Control/Monad/IOSim.hs]: Fair enough :-)

File: io-sim-classes/src/Control/Monad/Class/MonadFork.hs (hunk 10 above)
[the remark this answers is not in the document]
Comment (dcoutts) [reply io-sim-classes/src/Control/Monad/Class/MonadFork.hs]: I'm actually ok with having `ThreadId` and `myThreadId` in the `MonadFork` class. I think that makes sense independently. And extending `fork` to return the `ThreadId`. I might make a separate PR for this. I don't think we need `toThreadIdInt` though do we? If it's in `Eq`, and `Ord` and `Show`?

[the remark this answers is not in the document]
Comment (365andreas) [reply io-sim-classes/src/Control/Monad/Class/MonadFork.hs]: > I don't think we need `toThreadIdInt` though do we? If it's in `Eq`, and `Ord` and `Show`?
Yes I will remove it.

File: typed-protocols/src/Network/TypedProtocol/Driver.hs (hunk 8 above)
Comment (365andreas) [reply typed-protocols/src/Network/TypedProtocol/Driver.hs]: https://github.com/input-output-hk/ouroboros-network/blob/andreas/119-logging-in-Sim/typed-protocols/src/Network/TypedProtocol/Driver.hs#L101-L103

File: io-sim-classes/src/Control/Monad/Class/MonadProbe.hs [hunk 11]
```diff
@@ -60,3 +82,150 @@ instance MonadRunProbe IO IO where
...
+            { traceOutput = \a -> atomically $ modifyTVar' c (a:)
+            , readLogTrace = atomically $ do
+                v <- readTVar c
+                writeTVar c $ error "Read can only be called once!!"
+                return $ reverse v
+            , traceStopping = fail "Not supported operation"
+            }
+
+-- can call readLogTrace multiple times
+logTraceToSTMList' :: MonadSTM m => m (LogTrace [t] m t)
```
[the remark this answers is not in the document]
Comment (365andreas) [reply io-sim-classes/src/Control/Monad/Class/MonadProbe.hs]: Deleting it since calls to `fail` function are removed ant there is not any difference anymore.

File: typed-protocols/src/Network/TypedProtocol/PingPong/Tests.hs (hunk 9 above)
Comment (365andreas) [reply typed-protocols/src/Network/TypedProtocol/PingPong/Tests.hs]: No it was just forgotten there.

**Comment by 365andreas** [comment]
It depends on https://github.com/input-output-hk/ouroboros-network/pull/303.

File: io-sim-classes/src/Control/Monad/Class/MonadTimer.hs (hunk 2 above)
Comment (365andreas) [reply io-sim-classes/src/Control/Monad/Class/MonadTimer.hs]: @coot :arrow_up: Should we address this now?

**Comment by 365andreas** [comment]
@coot is there any `pkgs/generate.sh` to create automatically the nix files as it was in `cardano-sl`?

File: io-sim-classes/src/Control/Monad/Class/MonadProbe.hs (hunk 4 above)
Comment (365andreas) [reply io-sim-classes/src/Control/Monad/Class/MonadProbe.hs]: I cannot be in `ouroboros-network` since tests that use the logging are in `typed-protocols`.

**Comment by 365andreas** [comment]
@coot is there a `cabal2nix` command for `outoboros-network`? How I apply it?

File: io-sim-classes/src/Control/Monad/Class/MonadProbe.hs (hunk 4 above)
Comment (coot) [reply io-sim-classes/src/Control/Monad/Class/MonadProbe.hs]: You can also move the tests to `ouroboros-network`.

**Comment by coot** [comment]
@365andreas what do you want to do?  Update dependencies? That's manual work here.

File: io-sim-classes/src/Control/Monad/Class/MonadProbe.hs (hunk 4 above)
Comment (365andreas) [reply io-sim-classes/src/Control/Monad/Class/MonadProbe.hs]: I have added some logging messages  in `runPeer` which also lives in the lib of `typed-protocols`.

File: cabal.project [hunk 12]
```diff
@@ -1,5 +1,5 @@
 packages: ./typed-protocols
-	  ./ouroboros-network-testing
+	        ./ouroboros-network-testing
```
Comment (CodiePP) [inline cabal.project]: why?

**Comment by 365andreas** [comment]
Note: https://github.com/input-output-hk/ouroboros-network/pull/375 needs to be merged first.

File: ouroboros-consensus/src/Ouroboros/Consensus/Node.hs [hunk 13]
```diff
@@ -718,6 +721,7 @@ initNetworkLayer :: forall m up hdr blk.
                     , HasHeader hdr
                     , hdr ~ blk     -- TODO: for now
                     , Eq blk        -- TODO: for now
+                    , Show blk
```
Comment (avieth) [inline ouroboros-consensus/src/Ouroboros/Consensus/Node.hs]: Why do we need to add `Show blk`?

File: typed-protocols/src/Network/TypedProtocol/Driver.hs [hunk 14]
```diff
@@ -83,31 +85,37 @@ import Numeric.Natural (Natural)
 --
 runPeer
   :: forall ps (st :: ps) pr failure bytes m a .
-     (MonadThrow m, Exception failure)
-  => Codec ps failure m bytes
+     (MonadThrow m, Exception failure, Show (AnyMessage ps))
+  => Tracer m String
```
Comment (avieth) [inline typed-protocols/src/Network/TypedProtocol/Driver.hs]: Should use a different type IMO. A sum with each of the things that may be traced.

File: typed-protocols/src/Network/TypedProtocol/PingPong/Tests.hs [hunk 15]
```diff
@@ -269,10 +271,13 @@ prop_connect_pipelined5 choices (Positive omax) (NonNegative n) =
...
 prop_channel :: (MonadSTM m, MonadAsync m, MonadCatch m)
-             => NonNegative Int -> m Bool
-prop_channel (NonNegative n) = do
+             => NonNegative Int -> Tracer m String -> m Bool
+prop_channel (NonNegative n) tr = do
+    tracingWith tr "Starting ..."
     ((), n') <- runConnectedPeers createConnectedChannels
+                                  tr
                                   codecPingPong client server
+    tracingWith tr "... ending."
```
Comment (avieth) [inline typed-protocols/src/Network/TypedProtocol/PingPong/Tests.hs]: What's the usefulness of this and the `"Starting ..." trace?

Comment (365andreas) [reply typed-protocols/src/Network/TypedProtocol/PingPong/Tests.hs]: Just to indicate the start and the end of this process. Removed them for cleaner code.

File: ouroboros-consensus/src/Ouroboros/Consensus/Node.hs (hunk 13 above)
Comment (365andreas) [reply ouroboros-consensus/src/Ouroboros/Consensus/Node.hs]: If `sayTracer` (https://github.com/input-output-hk/ouroboros-network/pull/298/files#diff-a0e24f461e7fb4d85f24eba52bc9ae39R23) is used then we will have to set this constraint.

Comment (365andreas) [reply ouroboros-consensus/src/Ouroboros/Consensus/Node.hs]: Will remove this also.

File: typed-protocols/src/Network/TypedProtocol/Driver.hs (hunk 14 above)
Comment (365andreas) [reply typed-protocols/src/Network/TypedProtocol/Driver.hs]: What about `Tracer m Dynamic` ?

Comment (365andreas) [reply typed-protocols/src/Network/TypedProtocol/Driver.hs]: Or `Tracer m Traceable`
where 
```
data Traceable = Str String
                          | Blk Block
                          | Msg ...
```
?

Comment (365andreas) [reply typed-protocols/src/Network/TypedProtocol/Driver.hs]: https://github.com/input-output-hk/ouroboros-network/pull/298/commits/23b9bf52fefa676c5b1f44c1a6ec4fd69d617cea#diff-83b90bbd642add21a4351e069869662cR238

File: typed-protocols/src/Network/TypedProtocol/Codec.hs [hunk 16]
```diff
@@ -234,6 +235,9 @@ data AnyMessageAndAgency ps where
                       -> AnyMessageAndAgency ps
 
 
+data Traceable ps = Str String
```
[the remark this answers is not in the document]
Comment (coot) [reply typed-protocols/src/Network/TypedProtocol/Codec.hs]: We might need to hide `ps` as an existential, but you can leave it as such for now.

File: io-sim-classes/src/Control/Monad/Tracer.hs [hunk 17]
```diff
@@ -0,0 +1,32 @@
...
+import           Cardano.BM.Tracer (Tracer (..), nullTracer,
+                     tracingWith)
+
+import           Control.Monad.Class.MonadSay
+import           Control.Monad.Class.MonadSTM
+
+sayTracer :: (Show a, MonadSay m) => Tracer m a
+sayTracer = Tracer $ Op $ say . show
+
+toListTracer :: (MonadSTM m) => TVar m [a] -> Tracer m a
```
Comment (coot) [inline io-sim-classes/src/Control/Monad/Tracer.hs]: Maybe just `listTracer` and `dynListTracer`, when it will be used at call sites, it will look better I think.

File: cabal.project [hunk 18]
```diff
@@ -1,5 +1,5 @@
 packages: ./typed-protocols
-	  ./ouroboros-network-testing
+          ./ouroboros-network-testing
```
[the remark this answers is not in the document]
Comment (365andreas) [reply cabal.project]: I use only spaces.

[the remark this answers is not in the document]
Comment (365andreas) [reply cabal.project]: There was a tab (consisting of 2 spaces) in the start of the specific line.

File: typed-protocols/src/Network/TypedProtocol/PingPong/Tests.hs [hunk 19]
```diff
@@ -281,11 +285,21 @@ prop_channel (NonNegative n) = do
...
+        let (_msgs, bool) =
+                runSimOrThrow $ do
+                    c   <- newTVarM []
+                    res <- prop_channel n $ toListTracer c
+                    ms  <- atomically $ readTVar c
+                    return (ms, res)
+        -- forM_ (reverse _msgs) $ \case
+        --     Str m -> print m
+        --     Msg m -> print m
+        return $ property bool
```
Comment (365andreas) [inline typed-protocols/src/Network/TypedProtocol/PingPong/Tests.hs]: @coot With this change I wanted to demonstrate a use of `listTracer`. Should I revert these changes?

Comment (coot) [reply typed-protocols/src/Network/TypedProtocol/PingPong/Tests.hs]: Yes, I think you can remove it.

**Comment by 365andreas** [comment]
@coot I can squash the commits when it is 100% ready to be merged.

File: typed-protocols/src/Network/TypedProtocol/PingPong/Tests.hs (hunk 19 above)
Comment (dcoutts) [reply typed-protocols/src/Network/TypedProtocol/PingPong/Tests.hs]: I think for `SimM` directly we can use a more direct tracer, using the Sim's direct support of `Dynamic` in the trace output. It's like `MonadSay` but for any `Typeable`, and specific to the `SimM` monand, not with any class.

Comment (365andreas) [reply typed-protocols/src/Network/TypedProtocol/PingPong/Tests.hs]: @dcoutts Do you mean something like this?
```
dynTracer :: Typeable a => Tracer (SimM s) a
dynTracer = Tracer $ Op $ \a ->
                SimM $ \k -> Output (toDyn a) (k ())
```

**Comment by njd42** [comment]
> So I pruned this down to
> 
> ```
> data TraceSendRecv ps = TraceSendMsg (AnyMessage ps)
>                       | TraceRecvMsg (AnyMessage ps)
> ```

Agree about no *stringy* logging - also suggest that you might wish to record decisions in the peers, so additional constructors may be needed

**Comment by dcoutts** [comment]
> also suggest that you might wish to record decisions in the peers, so additional constructors may be needed
That'd be tracing in the block fetch logic, which is not covered yet. This is just a trace point in running every protocol.

File: typed-protocols/src/Network/TypedProtocol/PingPong/Tests.hs (hunk 19 above)
Comment (dcoutts) [reply typed-protocols/src/Network/TypedProtocol/PingPong/Tests.hs]: Yes.
